docs/language/learn-ql/ql-etudes/river-crossing.ql

```diff
@@ -0,0 +1,114 @@
+/** 
+ * @name River crossing puzzle
+ * @description This implements the classical puzzle where a man is trying to
+ * ferry a goat, a cabbage, and a wolf across a river.
+ * His boat can only take himself and at most one item as cargo.
+ * His problem is that if the goat is left alone with the cabbage,
+ * it will eat it.
+ * And if the wolf is left alone with the goat, it will eat it.
+ * How does he get everything across the river?
+ *
+ * Note: Parts of this QL file are included in the corresponding .rst file.
+ * Make sure to update the line numbers if you change anything here!
+ */
+
+/** A possible cargo item. */
+class Cargo extends string {
+  Cargo() {
+    this = "Nothing" or
+    this = "Goat" or
+    this = "Cabbage" or
+    this = "Wolf"
+  }
+}
+
+/** One of two shores. */
+class Shore extends string {
+  Shore() {
+    this = "Left" or
+    this = "Right"
+  }
+
+  /** Returns the other shore. */
+  Shore other() {
+    this = "Left" and result = "Right"
+    or
+    this = "Right" and result = "Left"
+  }
+}
+
+/** Renders the state as a string. */
+string renderState(Shore man, Shore goat, Shore cabbage, Shore wolf) {
+  result = man + "," + goat + "," + cabbage + "," + wolf
+}
+
+/** A record of where everything is. */
+class State extends string {
+  Shore man;
+
+  Shore goat;
+
+  Shore cabbage;
+
+  Shore wolf;
+
+  State() { this = renderState(man, goat, cabbage, wolf) }
+
+  /** Returns the state that is reached after ferrying a particular cargo item. */
+  State ferry(Cargo cargo) {
+    cargo = "Nothing" and result = renderState(man.other(), goat, cabbage, wolf)
+    or
+    cargo = "Goat" and result = renderState(man.other(), goat.other(), cabbage, wolf)
+    or
+    cargo = "Cabbage" and result = renderState(man.other(), goat, cabbage.other(), wolf)
+    or
+    cargo = "Wolf" and result = renderState(man.other(), goat, cabbage, wolf.other())
+  }
+
+  /**
+   * Holds if predator and prey are on the same shore and the man
+   * is not present.
+   */
+  predicate eats(Shore predator, Shore prey) { predator = prey and man != predator }
```

Comment:
I think this would also be clearer if you could rewrite the condition as `predatorLocation = preyLocation`.